lib/Db/FeedMapperV2.php

```diff
@@ -158,23 +160,34 @@ public function findAllFromFolder(?int $id): array
      * @param string   $userId
      * @param int      $id
      * @param int|null $maxItemID
+     *
+     * @throws \OCP\DB\Exception
+     *
+     * @TODO Update for NC 21
      */
     public function read(string $userId, int $id, ?int $maxItemID = null): void
     {
-        $builder = $this->db->getQueryBuilder();
-        $builder->update(ItemMapperV2::TABLE_NAME, 'items')
+        $idBuilder = $this->db->getQueryBuilder();
+        $idBuilder->select('items.id')
+            ->from(ItemMapperV2::TABLE_NAME, 'items')
             ->innerJoin('items', FeedMapperV2::TABLE_NAME, 'feeds', 'items.feed_id = feeds.id')
-            ->setValue('unread', 0)
             ->andWhere('feeds.user_id = :userId')
-            ->andWhere('feeds.id = :feedId')
-            ->setParameter('userId', $userId)
-            ->setParameter('feedId', $id);
+            ->andWhere('feeds.id = :feedId');
+
+        $idList = $this->db->executeQuery($idBuilder->getSQL(), $idBuilder->getParameters())
+                           ->fetch(PDO::FETCH_COLUMN);
+
+        $builder = $this->db->getQueryBuilder();
+        $builder->update(ItemMapperV2::TABLE_NAME)
+            ->set('unread', $builder->createParameter('unread'))
+            ->where($builder->expr()->in('id', $idList))
```

Comment:
The `$idList` is somehow interpreted/escaped as column names instead of values inside the generated `IN ( )`.
Maybe you could also set it using a parameter and the explicit type `PARAM_INT_ARRAY`.

<details><summary>Related log entry</summary>

```json
{
    "reqId": "8MlXh83z8aMDOg0NGlgN",
    "level": 3,
    "time": "2021-03-20T17:16:48+00:00",
    "remoteAddr": "127.0.0.1",
    "user": "anoymouserver",
    "app": "index",
    "method": "POST",
    "url": "/apps/news/feeds/135/read",
    "message": {
        "Exception": "Doctrine\\DBAL\\Exception\\InvalidFieldNameException",
        "Message": "An exception occurred while executing 'UPDATE `oc_news_items` SET `unread` = ? WHERE (`id` IN (`107687`)) AND (id <= ?)' with params [false, 129183]:\n\nSQLSTATE[42S22]: Column not found: 1054 Unknown column '107687' in 'where clause'",
        "Code": 0,
        "Trace": [{
                "file": "/var/www/cloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php",
                "line": 169,
                "function": "convertException",
                "class": "Doctrine\\DBAL\\Driver\\AbstractMySQLDriver",
                "type": "->",
                "args": ["An exception occurred while executing 'UPDATE `oc_news_items` SET `unread` = ? WHERE (`id` IN (`107687`)) AND (id <= ?)' with params [false, 129183]:\n\nSQLSTATE[42S22]: Column not found: 1054 Unknown column '107687' in 'where clause'", {
                        "errorInfo": ["42S22", 1054, "Unknown column '107687' in 'where clause'"],
                        "__class__": "Doctrine\\DBAL\\Driver\\PDOException"
                    }
                ]
            }, {
                "file": "/var/www/cloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php",
                "line": 149,
                "function": "wrapException",
                "class": "Doctrine\\DBAL\\DBALException",
                "type": "::",
                "args": [{
                        "__class__": "Doctrine\\DBAL\\Driver\\PDOMySql\\Driver"
                    }, {
                        "errorInfo": ["42S22", 1054, "Unknown column '107687' in 'where clause'"],
                        "__class__": "Doctrine\\DBAL\\Driver\\PDOException"
                    }, "An exception occurred while executing 'UPDATE `oc_news_items` SET `unread` = ? WHERE (`id` IN (`107687`)) AND (id <= ?)' with params [false, 129183]:\n\nSQLSTATE[42S22]: Column not found: 1054 Unknown column '107687' in 'where clause'"]
            }, {
                "file": "/var/www/cloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php",
                "line": 1071,
                "function": "driverExceptionDuringQuery",
                "class": "Doctrine\\DBAL\\DBALException",
                "type": "::",
                "args": [{
                        "__class__": "Doctrine\\DBAL\\Driver\\PDOMySql\\Driver"
                    }, {
                        "errorInfo": ["42S22", 1054, "Unknown column '107687' in 'where clause'"],
                        "__class__": "Doctrine\\DBAL\\Driver\\PDOException"
                    }, "UPDATE `oc_news_items` SET `unread` = ? WHERE (`id` IN (`107687`)) AND (id <= ?)", {
                        "1": false,
                        "2": 129183
                    }
                ]
            }, {
                "file": "/var/www/cloud/lib/private/DB/Connection.php",
                "line": 222,
                "function": "executeUpdate",
                "class": "Doctrine\\DBAL\\Connection",
                "type": "->",
                "args": ["UPDATE `oc_news_items` SET `unread` = ? WHERE (`id` IN (`107687`)) AND (id <= ?)", [false, 129183], [2, 2]]
            }, {
                "file": "/var/www/cloud/apps/news/lib/Db/FeedMapperV2.php",
                "line": 193,
                "function": "executeUpdate",
                "class": "OC\\DB\\Connection",
                "type": "->",
                "args": ["UPDATE `oc_news_items` SET `unread` = :unread WHERE (`id` IN (`107687`)) AND (id <= :maxItemId)", {
                        "unread": false,
                        "maxItemId": 129183
                    }
                ]
            }, {
                "file": "/var/www/cloud/apps/news/lib/Service/FeedServiceV2.php",
                "line": 358,
                "function": "read",
                "class": "OCA\\News\\Db\\FeedMapperV2",
                "type": "->",
                "args": ["anoymouserver", 135, 129183]
            }, {
                "file": "/var/www/cloud/apps/news/lib/Controller/FeedController.php",
                "line": 286,
                "function": "read",
                "class": "OCA\\News\\Service\\FeedServiceV2",
                "type": "->",
                "args": ["anoymouserver", 135, 129183]
            }, {
                "file": "/var/www/cloud/lib/private/AppFramework/Http/Dispatcher.php",
                "line": 169,
                "function": "read",
                "class": "OCA\\News\\Controller\\FeedController",
                "type": "->",
                "args": [135, 129183]
            }, {
                "file": "/var/www/cloud/lib/private/AppFramework/Http/Dispatcher.php",
                "line": 100,
                "function": "executeController",
                "class": "OC\\AppFramework\\Http\\Dispatcher",
                "type": "->",
                "args": [{
                        "__class__": "OCA\\News\\Controller\\FeedController"
                    }, "read"]
            }, {
                "file": "/var/www/cloud/lib/private/AppFramework/App.php",
                "line": 152,
                "function": "dispatch",
                "class": "OC\\AppFramework\\Http\\Dispatcher",
                "type": "->",
                "args": [{
                        "__class__": "OCA\\News\\Controller\\FeedController"
                    }, "read"]
            }, {
                "file": "/var/www/cloud/lib/private/Route/Router.php",
                "line": 309,
                "function": "main",
                "class": "OC\\AppFramework\\App",
                "type": "::",
                "args": ["OCA\\News\\Controller\\FeedController", "read", {
                        "__class__": "OC\\AppFramework\\DependencyInjection\\DIContainer"
                    }, {
                        "feedId": "135",
                        "_route": "news.feed.read"
                    }
                ]
            }, {
                "file": "/var/www/cloud/lib/base.php",
                "line": 1008,
                "function": "match",
                "class": "OC\\Route\\Router",
                "type": "->",
                "args": ["/apps/news/feeds/135/read"]
            }, {
                "file": "/var/www/cloud/index.php",
                "line": 37,
                "function": "handleRequest",
                "class": "OC",
                "type": "::",
                "args": []
            }
        ],
        "File": "/var/www/cloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php",
        "Line": 60,
        "Previous": {
            "Exception": "Doctrine\\DBAL\\Driver\\PDOException",
            "Message": "SQLSTATE[42S22]: Column not found: 1054 Unknown column '107687' in 'where clause'",
            "Code": "42S22",
            "Trace": [{
                    "file": "/var/www/cloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php",
                    "line": 1062,
                    "function": "execute",
                    "class": "Doctrine\\DBAL\\Driver\\PDOStatement",
                    "type": "->",
                    "args": []
                }, {
                    "file": "/var/www/cloud/lib/private/DB/Connection.php",
                    "line": 222,
                    "function": "executeUpdate",
                    "class": "Doctrine\\DBAL\\Connection",
                    "type": "->",
                    "args": ["UPDATE `oc_news_items` SET `unread` = ? WHERE (`id` IN (`107687`)) AND (id <= ?)", [false, 129183], [2, 2]]
                }, {
                    "file": "/var/www/cloud/apps/news/lib/Db/FeedMapperV2.php",
                    "line": 193,
                    "function": "executeUpdate",
                    "class": "OC\\DB\\Connection",
                    "type": "->",
                    "args": ["UPDATE `oc_news_items` SET `unread` = :unread WHERE (`id` IN (`107687`)) AND (id <= :maxItemId)", {
                            "unread": false,
                            "maxItemId": 129183
                        }
                    ]
                }, {
                    "file": "/var/www/cloud/apps/news/lib/Service/FeedServiceV2.php",
                    "line": 358,
                    "function": "read",
                    "class": "OCA\\News\\Db\\FeedMapperV2",
                    "type": "->",
                    "args": ["anoymouserver", 135, 129183]
                }, {
                    "file": "/var/www/cloud/apps/news/lib/Controller/FeedController.php",
                    "line": 286,
                    "function": "read",
                    "class": "OCA\\News\\Service\\FeedServiceV2",
                    "type": "->",
                    "args": ["anoymouserver", 135, 129183]
                }, {
                    "file": "/var/www/cloud/lib/private/AppFramework/Http/Dispatcher.php",
                    "line": 169,
                    "function": "read",
                    "class": "OCA\\News\\Controller\\FeedController",
                    "type": "->",
                    "args": [135, 129183]
                }, {
                    "file": "/var/www/cloud/lib/private/AppFramework/Http/Dispatcher.php",
                    "line": 100,
                    "function": "executeController",
                    "class": "OC\\AppFramework\\Http\\Dispatcher",
                    "type": "->",
                    "args": [{
                            "__class__": "OCA\\News\\Controller\\FeedController"
                        }, "read"]
                }, {
                    "file": "/var/www/cloud/lib/private/AppFramework/App.php",
                    "line": 152,
                    "function": "dispatch",
                    "class": "OC\\AppFramework\\Http\\Dispatcher",
                    "type": "->",
                    "args": [{
                            "__class__": "OCA\\News\\Controller\\FeedController"
                        }, "read"]
                }, {
                    "file": "/var/www/cloud/lib/private/Route/Router.php",
                    "line": 309,
                    "function": "main",
                    "class": "OC\\AppFramework\\App",
                    "type": "::",
                    "args": ["OCA\\News\\Controller\\FeedController", "read", {
                            "__class__": "OC\\AppFramework\\DependencyInjection\\DIContainer"
                        }, {
                            "feedId": "135",
                            "_route": "news.feed.read"
                        }
                    ]
                }, {
                    "file": "/var/www/cloud/lib/base.php",
                    "line": 1008,
                    "function": "match",
                    "class": "OC\\Route\\Router",
                    "type": "->",
                    "args": ["/apps/news/feeds/135/read"]
                }, {
                    "file": "/var/www/cloud/index.php",
                    "line": 37,
                    "function": "handleRequest",
                    "class": "OC",
                    "type": "::",
                    "args": []
                }
            ],
            "File": "/var/www/cloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php",
            "Line": 129,
            "Previous": {
                "Exception": "PDOException",
                "Message": "SQLSTATE[42S22]: Column not found: 1054 Unknown column '107687' in 'where clause'",
                "Code": "42S22",
                "Trace": [{
                        "file": "/var/www/cloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php",
                        "line": 127,
                        "function": "execute",
                        "class": "PDOStatement",
                        "type": "->",
                        "args": [null]
                    }, {
                        "file": "/var/www/cloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php",
                        "line": 1062,
                        "function": "execute",
                        "class": "Doctrine\\DBAL\\Driver\\PDOStatement",
                        "type": "->",
                        "args": []
                    }, {
                        "file": "/var/www/cloud/lib/private/DB/Connection.php",
                        "line": 222,
                        "function": "executeUpdate",
                        "class": "Doctrine\\DBAL\\Connection",
                        "type": "->",
                        "args": ["UPDATE `oc_news_items` SET `unread` = ? WHERE (`id` IN (`107687`)) AND (id <= ?)", [false, 129183], [2, 2]]
                    }, {
                        "file": "/var/www/cloud/apps/news/lib/Db/FeedMapperV2.php",
                        "line": 193,
                        "function": "executeUpdate",
                        "class": "OC\\DB\\Connection",
                        "type": "->",
                        "args": ["UPDATE `oc_news_items` SET `unread` = :unread WHERE (`id` IN (`107687`)) AND (id <= :maxItemId)", {
                                "unread": false,
                                "maxItemId": 129183
                            }
                        ]
                    }, {
                        "file": "/var/www/cloud/apps/news/lib/Service/FeedServiceV2.php",
                        "line": 358,
                        "function": "read",
                        "class": "OCA\\News\\Db\\FeedMapperV2",
                        "type": "->",
                        "args": ["anoymouserver", 135, 129183]
                    }, {
                        "file": "/var/www/cloud/apps/news/lib/Controller/FeedController.php",
                        "line": 286,
                        "function": "read",
                        "class": "OCA\\News\\Service\\FeedServiceV2",
                        "type": "->",
                        "args": ["anoymouserver", 135, 129183]
                    }, {
                        "file": "/var/www/cloud/lib/private/AppFramework/Http/Dispatcher.php",
                        "line": 169,
                        "function": "read",
                        "class": "OCA\\News\\Controller\\FeedController",
                        "type": "->",
                        "args": [135, 129183]
                    }, {
                        "file": "/var/www/cloud/lib/private/AppFramework/Http/Dispatcher.php",
                        "line": 100,
                        "function": "executeController",
                        "class": "OC\\AppFramework\\Http\\Dispatcher",
                        "type": "->",
                        "args": [{
                                "__class__": "OCA\\News\\Controller\\FeedController"
                            }, "read"]
                    }, {
                        "file": "/var/www/cloud/lib/private/AppFramework/App.php",
                        "line": 152,
                        "function": "dispatch",
                        "class": "OC\\AppFramework\\Http\\Dispatcher",
                        "type": "->",
                        "args": [{
                                "__class__": "OCA\\News\\Controller\\FeedController"
                            }, "read"]
                    }, {
                        "file": "/var/www/cloud/lib/private/Route/Router.php",
                        "line": 309,
                        "function": "main",
                        "class": "OC\\AppFramework\\App",
                        "type": "::",
                        "args": ["OCA\\News\\Controller\\FeedController", "read", {
                                "__class__": "OC\\AppFramework\\DependencyInjection\\DIContainer"
                            }, {
                                "feedId": "135",
                                "_route": "news.feed.read"
                            }
                        ]
                    }, {
                        "file": "/var/www/cloud/lib/base.php",
                        "line": 1008,
                        "function": "match",
                        "class": "OC\\Route\\Router",
                        "type": "->",
                        "args": ["/apps/news/feeds/135/read"]
                    }, {
                        "file": "/var/www/cloud/index.php",
                        "line": 37,
                        "function": "handleRequest",
                        "class": "OC",
                        "type": "::",
                        "args": []
                    }
                ],
                "File": "/var/www/cloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php",
                "Line": 127
            }
        },
        "CustomMessage": "--"
    },
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:86.0) Gecko/20100101 Firefox/86.0",
    "version": "20.0.8.1",
    "id": "60562e01100a9"
}

```
</details>